src/replication.c

```diff
@@ -2249,25 +2330,32 @@ void readSyncBulkPayload(connection *conn) {
         }
 
         zfree(server.repl_transfer_tmpfile);
-        server.repl_transfer_fd = -1;
+            server.repl_transfer_fd = -1;
         server.repl_transfer_tmpfile = NULL;
     }
 
     /* Final setup of the connected slave <- master link */
-    replicationCreateMasterClient(server.repl_transfer_s,rsi.repl_stream_db);
-    server.repl_state = REPL_STATE_CONNECTED;
+    if (isRdbConnectionSync(conn)) {
+        /* In case of fsync using rdb channel, the master client was already created for psync purposes
+         * Instead of creating a new client we will use the one created for psync */
+        updateReplicationStateActive(conn);
+    } else {
+        server.master = replicationCreateMasterClient(server.repl_transfer_s, rsi.repl_stream_db);
+        server.repl_state = REPL_STATE_CONNECTED;
+    }
     server.repl_down_since = 0;
 
     /* Fire the master link modules event. */
     moduleFireServerEvent(REDISMODULE_EVENT_MASTER_LINK_CHANGE,
                           REDISMODULE_SUBEVENT_MASTER_LINK_UP,
                           NULL);
-
-    /* After a full resynchronization we use the replication ID and
-     * offset of the master. The secondary ID / offset are cleared since
-     * we are starting a new history. */
-    memcpy(server.replid,server.master->replid,sizeof(server.replid));
-    server.master_repl_offset = server.master->reploff;
+    if (server.repl_state == REPL_STATE_CONNECTED) {
+        /* After a full resynchronization we use the replication ID and
+        * offset of the master. The secondary ID / offset are cleared since
+        * we are starting a new history. */
+        memcpy(server.replid,server.master->replid,sizeof(server.replid));
+        server.master_repl_offset = server.master->reploff;
+    }
```

Comment:
I did some refactoring, I think that after that the flow now will be clearer. 
After refactor, in the race condition case (RDB is loaded before psync established), main connection will call 

-> `slaveTryPartialResynchronization`
-> `completeTaskRDBChannelSync` 
-> `replicationResurrectProvisionalMaster` 
  there it will initialize `server.master` with the relevant fields.
  * note that some of the method has been renamed